cpp/velox/shuffle/VeloxSplitter.cc

```diff
@@ -0,0 +1,1360 @@
+#include "VeloxSplitter.h"
+#include "VeloxSplitterPartitionWriter.h"
+#include "velox/vector/arrow/Bridge.h"
+
+#include "utils/compression.h"
+#include "utils/macros.h"
+
+#include "include/arrow/c/bridge.h"
+
+#if defined(__x86_64__)
+#include <immintrin.h>
+#include <x86intrin.h>
+#elif defined(__aarch64__)
+#include <arm_neon.h>
+#endif
+
+#include <iostream>
+
+using namespace facebook;
+
+#ifndef SPLIT_BUFFER_SIZE
+// by default, allocate 8M block, 2M page size
+#define SPLIT_BUFFER_SIZE 16 * 1024 * 1024
+#endif
+
+namespace gluten {
+
+// VeloxSplitter
+arrow::Result<std::shared_ptr<VeloxSplitter>>
+VeloxSplitter::Make(const std::string& name, uint32_t num_partitions, SplitOptions options) {
+  std::shared_ptr<VeloxSplitter> splitter = nullptr;
+  if (name == "hash") {
+    splitter = VeloxSplitter::Create<VeloxHashSplitter>(num_partitions, options);
+  } else if (name == "rr") {
+    splitter = VeloxSplitter::Create<VeloxRoundRobinSplitter>(num_partitions, options);
+  } else if (name == "range") {
+    splitter = VeloxSplitter::Create<VeloxFallbackRangeSplitter>(num_partitions, options);
+  } else if (name == "single") {
+    splitter = VeloxSplitter::Create<VeloxSinglePartSplitter>(num_partitions, options);
+  }
+
+  if (splitter) {
+    return splitter->Init();
+  }
+  return arrow::Status::NotImplemented("Partitioning " + name + " not supported yet.");
+}
+
+arrow::Status VeloxSplitter::Init() {
+#if defined(__x86_64__)
+  support_avx512_ = __builtin_cpu_supports("avx512bw");
+#else
+  support_avx512_ = false;
+#endif
+
+  // partition number should be less than 64k
+  ARROW_CHECK_LE(num_partitions_, 64 * 1024);
```

Comment:
why num_partitions_  can not be greater then 64K?

Reply:
Because the partition id is 16bit to save memory footprint. If your case needs more than 64K reducers, feel free to add 32bit support through template.